internal/v5/api.go

```diff
@@ -1107,21 +1107,66 @@ func (h *ReqHandler) serveMacaroon(_ http.Header, _ *http.Request) (interface{},
 // GET /delegatable-macaroon
 // See https://github.com/juju/charmstore/blob/v4/docs/API.md#get-delegatable-macaroon
 func (h *ReqHandler) serveDelegatableMacaroon(_ http.Header, req *http.Request) (interface{}, error) {
+	values, err := url.ParseQuery(req.URL.RawQuery)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	entityIds := values["id"]
+	if len(entityIds) == 0 {
+		auth, err := h.authorize(req, []string{params.Everyone}, true, nil)
+		if err != nil {
+			return nil, errgo.Mask(err, errgo.Any)
+		}
+		if auth.Username == "" {
+			return nil, errgo.WithCausef(nil, params.ErrForbidden, "delegatable macaroon is not obtainable using admin credentials")
+		}
+		// TODO propagate expiry time from macaroons in request.
+		m, err := h.Store.Bakery.NewMacaroon("", nil, []checkers.Caveat{
+			checkers.DeclaredCaveat(UsernameAttr, auth.Username),
+			checkers.TimeBeforeCaveat(time.Now().Add(DelegatableMacaroonExpiry)),
+			checkers.DenyCaveat(opAccessCharmWithTerms),
+		})
+		if err != nil {
+			return nil, errgo.Mask(err)
+		}
+		return m, nil
+	}
+	var resolvedURLs []*router.ResolvedURL
```

Comment:
resolvedURLs := make([]*router.ResolvedURL, len(entityIds))

to save a few allocations?